x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/searchablesnapshots/SearchableSnapshotStats.java

```diff
@@ -0,0 +1,397 @@
+                    builder.endObject();
+                }
+                {
+                    builder.startObject("backward_seeks");
+                    builder.field("small", getBackwardSmallSeeks());
+                    builder.field("large", getBackwardLargeSeeks());
+                    builder.endObject();
+                }
+            }
+            return builder.endObject();
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            }
+            if (other == null || getClass() != other.getClass()) {
+                return false;
+            }
+            CacheIndexInputStats stats = (CacheIndexInputStats) other;
+            return fileLength == stats.fileLength
+                && openCount == stats.openCount
+                && innerCount == stats.innerCount
+                && closeCount == stats.closeCount
+                && Objects.equals(fileName, stats.fileName)
+                && Objects.equals(forwardSmallSeeks, stats.forwardSmallSeeks)
+                && Objects.equals(backwardSmallSeeks, stats.backwardSmallSeeks)
+                && Objects.equals(forwardLargeSeeks, stats.forwardLargeSeeks)
+                && Objects.equals(backwardLargeSeeks, stats.backwardLargeSeeks)
+                && Objects.equals(contiguousReads, stats.contiguousReads)
+                && Objects.equals(nonContiguousReads, stats.nonContiguousReads)
+                && Objects.equals(cachedBytesRead, stats.cachedBytesRead)
+                && Objects.equals(cachedBytesWritten, stats.cachedBytesWritten)
+                && Objects.equals(directBytesRead, stats.directBytesRead);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(fileName, fileLength, openCount, innerCount, closeCount,
+                forwardSmallSeeks, backwardSmallSeeks,
+                forwardLargeSeeks, backwardLargeSeeks,
+                contiguousReads, nonContiguousReads,
+                cachedBytesRead, cachedBytesWritten,
+                directBytesRead);
+        }
+    }
+
+    public static class Counter implements Writeable, ToXContentObject {
+
+        private final long count;
+        private final long total;
+        private final long min;
+        private final long max;
+
+        public Counter(final long count, final long total, final long min, final long max) {
+            this.count = count;
+            this.total = total;
+            this.min = min;
+            this.max = max;
+        }
+
+        Counter(final StreamInput in) throws IOException {
+            this.count = in.readZLong();
+            this.total = in.readZLong();
+            this.min = in.readZLong();
+            this.max = in.readZLong();
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeZLong(count);
```

Comment:
Should we use `writeVLong`? I think all these numbers will be nonnegative (we can assert that).

Reply:
They can be negative for backward seeks counters, which kind of makes sense. I also think we should handle negative longs here because `Counter` might have different purpose and count different things.

Anyway I pushed b1d79735f562188b5a0167b7c4400dff0f36c61e to use negative longs in `SearchableSnapshotShardStats` serializing tests and also to add some randomization for the seeking threshold in `IndexInputStatsTests`.

Reply:
Ah of course. :+1: